internal/ot/naor_pinkas.go

```diff
@@ -0,0 +1,170 @@
+package ot
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/optable/match/internal/crypto"
+	"github.com/optable/match/internal/util"
+)
+
+/*
+1 out of 2 base OT
+from the paper: Efficient Oblivious Transfer Protocol
+by Moni Naor and Benny Pinkas in 2001.
+reference: https://dl.acm.org/doi/abs/10.5555/365411.365502
+*/
+
+type naorPinkas struct {
+	baseCount int
+	msgLen    []int
+}
+
+func newNaorPinkas(baseCount int, msgLen []int) (naorPinkas, error) {
+	if len(msgLen) != baseCount {
+		return naorPinkas{}, ErrBaseCountMissMatch
+	}
+	return naorPinkas{baseCount: baseCount, msgLen: msgLen}, nil
+}
+
+func (n naorPinkas) Send(messages [][][]byte, rw io.ReadWriter) (err error) {
+	if len(messages) != n.baseCount {
+		return ErrBaseCountMissMatch
+	}
+
+	// Instantiate Reader, Writer
+	reader := crypto.NewECPointReader(rw)
+	writer := crypto.NewECPointWriter(rw)
+
+	// generate sender point A w/o secret, since a is never used.
+	_, pointA, err := crypto.GenerateKeyWithPoints()
+	if err != nil {
+		return err
+	}
+
+	// generate sender secret public key pairs used for encryption.
+	secretR, pointR, err := crypto.GenerateKeyWithPoints()
+	if err != nil {
+		return err
+	}
+
+	// send point A to receiver
+	if err := writer.Write(pointA); err != nil {
+		return err
+	}
+
+	// send point R to receiver
+	if err := writer.Write(pointR); err != nil {
+		return err
+	}
+
+	// precompute A = rA
+	pointA = pointA.ScalarMult(secretR)
+
+	// make a slice of points to receive K0.
+	pointK0 := make([]*crypto.Point, n.baseCount)
+	for i := range pointK0 {
+		pointK0[i] = crypto.NewPoint()
+		if err := reader.Read(pointK0[i]); err != nil {
+			return err
+		}
+	}
+
+	pointK := make([]*crypto.Point, 2)
+	var ciphertext []byte
+	// encrypt plaintext messages and send them.
+	for i := 0; i < n.baseCount; i++ {
+		// compute K0 = rK0
+		pointK[0] = pointK0[i].ScalarMult(secretR)
+		// compute K1 = rA - rK0
+		pointK[1] = pointA.Sub(pointK[0])
+
+		// encrypt plaintext message with key derived from K0, K1
+		for choice, plaintext := range messages[i] {
```

Comment:
I had to reread this several times. We should really document what messages is. Here are a few suggestions.

1. make it a [][2][]byte to enforce the two choices and document that it is a list of message choices.
2. make it a []OTMessageChoice and have OTMessageChoice simply contain a [2][]byte. It would put a name on the concept.